enhancements/network/ovs-dynamic-cpu-pinning.md

```diff
@@ -0,0 +1,205 @@
+  disable the feature cluster-wide.
+
+```mermaid
+sequenceDiagram
+    Note right of NodeTuningOperator: via MachineConfig
+  
+    NodeTuningOperator->>ovs-vswitchd: set_slice
+    NodeTuningOperator->>ovsdb-server: set_slice
+    NodeTuningOperator->>/etc/openvswitch/enable_cpu_pinning: fwrite('1')
+    activate /etc/openvswitch/enable_cpu_pinning
+    ovnkube-node ->> /etc/openvswitch/enable_cpu_pinning: fstat()
+    
+    activate ovnkube-node
+    loop
+    ovnkube-node ->> ovnkube-node: x = sched_getaffinity(x)
+    ovnkube-node ->> ovs-vswitchd: sched_setaffinity(x)
+    ovnkube-node ->> ovsdb-server: sched_setaffinity(x)
+    end
+
+    deactivate ovnkube-node
+    deactivate /etc/openvswitch/enable_cpu_pinning
+```
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+#### CPU aligning
+
+CPU alignment is implemented using system calls SYS_SCHED_SETAFFINITY and SYS_SCHED_GETAFFINITY.
+Pods from `ovnkube-node` daemonset run in the Burstable Quality of Service (QoS) class, hence their CPU 
+affinities are managed by the [kubelet cpu manager](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/).
+
+#### `/etc/openvswitchd/enable_cpu_pinning` file
+
+The file is used to activate the feature on ovnkube-node according to the following logic:
+a. If the file is missing OR the file file is empty, then the feature is turned `OFF`.
+b. If the file exists and its content is not an empty string, then the feature is turned `ON`.
+c. `ovnkube-node` responds to the file modifications (deletion or content update)
+  at runtime, without the need for a restart.
+
+#### Must-Gather enhancement
+
+As the feature is enabled via a fail on the host filesystem, it's useful to
+add `/etc/openvswitchd/enable_cpu_pinning` file to the [gather_network_logs](https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_network_logs) script.
+
+### Risks and Mitigations
+
+|Risk|Mitigation|
+|---|---|
+|After NTO enabled the feature, a network performance decrease occurred in some cluster nodes.|The cluster administrator can turn off the feature on specific nodes by connecting through SSH and deleting the activation file. An overriding MachineConfig can be created to achieve the same goal.|
+
+
+### Drawbacks
+
+This enhancement breaks the hard division between `kube-reserved` and `system-reserved` CPU pools in a cluster with `cpuManagerPolicy = static`.
+
+
+## Design Details
+
+### Test Plan
+
+Following scenario should be covered by automated tests:
+
+_Given_ an OpenShift cluster with a node N with 10 CPUs (0,1,...,9), configured with 
+- `cpuManagerPolicy = static`
+- `system-reserved = 0,1,2,3`
+
+_And_ a running pod with `Guaranteed` QoS running on cpu 4,5
+
+_When_ the administrator creates a the file `/etc/openvswitch/enable_cpu_pinning` on N's host filesystem
+
+_Then_ `ovs-switchd` and `ovsdb-server` processes on N has CPU affinity equal to `6,7,8,9`.
```

Comment:
It will have 0-3 as well.

Reply:
Well, It can be a little bit difficult from the ovnkube-node perspective. The only two ways it can be implemented that come to my mind are:
a. `ovnkube-node` knows the set of system-reserved CPUs, but I don't know how it can get that information.
b. `ovnkube-node` saves the cpuset which ovs starts with in order to add its mask to that cpuset dynamically. It should save that information on the host filesystem to make it work after the pod reboots, making the implementation much more tricky. 

Isn't it enough to make ovs run on kube cpus?

Reply:
Burstable pods have all cpus in their mask, including reserved. Nothing complicated to do here.

Reply:
Great, I didn't know that. Thanks for explaining